lib/Errors.ml

```diff
@@ -3,65 +3,135 @@
+          ; colnum= pos_cnum - pos_bol
+          ; included_from=
+              ( match fnames with
+              | [] -> None
+              | fnames :: _ -> Some (parse_location_from_string fnames) ) } )
+
+(** Take the AST.location_span corresponding to a pair of Lexing.position's *)
+let location_span_of_pos start_pos end_pos =
+  { start_loc= location_of_position start_pos
+  ; end_loc= location_of_position end_pos }
+
+(** Render a location as a string *)
+let rec create_string_from_location loc =
+  let open Format in
+  let included_from_str =
+    match loc.included_from with
+    | None -> ""
+    | Some loc2 ->
+        sprintf ", included from\n%s" (create_string_from_location loc2)
+  in
+  sprintf "file %s, line %d, column %d%s" loc.filename loc.linenum loc.colnum
+    included_from_str
+
+(** Render a location_span as a string *)
+let create_string_from_location_span loc_sp =
+  match loc_sp with {start_loc; end_loc} ->
+    let open Format in
+    let filename_str = sprintf "file %s, " start_loc.filename in
+    let linenum_str =
+      if end_loc.linenum = start_loc.linenum then
+        sprintf "line %d, " end_loc.linenum
+      else sprintf "lines %d-%d, " start_loc.linenum end_loc.linenum
     in
+    let colnum_str =
+      if end_loc.linenum = start_loc.linenum then
+        sprintf "columns %d-%d" start_loc.colnum end_loc.colnum
+      else sprintf "column %d" start_loc.colnum
+    in
+    let included_from_str =
+      match start_loc.included_from with
+      | None -> ""
+      | Some loc ->
+          sprintf ", included from\n%s" (create_string_from_location loc)
+    in
+    sprintf "%s%s%s%s" filename_str linenum_str colnum_str included_from_str
+
+(** We quote the lines before and after a certain location in the program *)
+let print_context {filename; linenum; colnum; _} =
+  try
     let open In_channel in
-    let input = create bare_file in
-    for _ = 1 to line - 3 do
+    let input = create filename in
+    for _ = 1 to linenum - 3 do
       ignore (input_line_exn input)
     done ;
     let open Printf in
     let line_2_before =
-      if line > 2 then sprintf "%6d:  %s\n" (line - 2) (input_line_exn input)
+      if linenum > 2 then
+        sprintf "%6d:  %s\n" (linenum - 2) (input_line_exn input)
       else ""
     in
     let line_before =
-      if line > 1 then sprintf "%6d:  %s\n" (line - 1) (input_line_exn input)
+      if linenum > 1 then
+        sprintf "%6d:  %s\n" (linenum - 1) (input_line_exn input)
       else ""
     in
-    let our_line = sprintf "%6d:  %s\n" line (input_line_exn input) in
-    let cursor_line = String.make (column + 9) ' ' ^ "^\n" in
+    let our_line = sprintf "%6d:  %s\n" linenum (input_line_exn input) in
+    let cursor_line = String.make (colnum + 9) ' ' ^ "^\n" in
     let line_after =
-      try sprintf "%6d:  %s\n" (line + 1) (input_line_exn input) with _ -> ""
+      try sprintf "%6d:  %s\n" (linenum + 1) (input_line_exn input) with _ ->
```

Comment:
Sure. It doesn't seem to be much of an efficiency concern, but people in OCaml still seem to prefer using option instead of error handling because it's more type safe: https://stackoverflow.com/questions/12160390/ocaml-performance-of-exceptions .
In this case, we're in a situation where we're handling exceptions already (e.g. a parsing error) so I felt it wasn't so bad, but I've changed it.